tests/unit/s2n_certificate_parsing_test.c

```diff
@@ -0,0 +1,445 @@
+#include "utils/s2n_safety.h"
+
+#define CERTIFICATE_1                                                  \
+    "MIIBljCCATygAwIBAgIUHxKbtYzLM4Bct5v5Sagb8aZU/BcwCgYIKoZIzj0EAwIw" \
+    "HjELMAkGA1UEBhMCVVMxDzANBgNVBAMMBmJyYW5jaDAgFw0yNDAxMjMwMDU3MTha" \
+    "GA8yMjAzMDYzMDAwNTcxOFowHDELMAkGA1UEBhMCVVMxDTALBgNVBAMMBGxlYWYw" \
+    "WTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASAmJKXt9P8hSz9ubntEokFn06+Rexr" \
+    "lEujwWUIq5Kl6QwvtfDCzrJN/sUmM5mssjq7pF6XVr6zFcQ6G4BfnwGio1gwVjAU" \
+    "BgNVHREEDTALgglsb2NhbGhvc3QwHQYDVR0OBBYEFNlOWJn7XfzC6xORzzjrdnqK" \
+    "UlJwMB8GA1UdIwQYMBaAFFkQhpDCJ3bAbXw1tXpmk5Fi7YIGMAoGCCqGSM49BAMC" \
+    "A0gAMEUCIB58OBwIruTJIy1f3tUgM/wXoO7fCoU25sMcioHBV9dYAiEA7Ufxa2JF" \
+    "I5LP6RGyllsjjnh0MJy1ZMXhw7X6GqFn4Rw="
+
+#define CERTIFICATE_2                                                  \
+    "MIIBoDCCAUegAwIBAgIUQud1+tNUPAEBnJLb2Lyl3/vuA4EwCgYIKoZIzj0EAwIw" \
+    "HDELMAkGA1UEBhMCVVMxDTALBgNVBAMMBHJvb3QwIBcNMjQwMTIzMDA1NzE4WhgP" \
+    "MjIwMzA2MzAwMDU3MThaMB4xCzAJBgNVBAYTAlVTMQ8wDQYDVQQDDAZicmFuY2gw" \
+    "WTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATfScB9w/LkHBAVXiyKN941555oyBpv" \
+    "IZeCNXX+gbSvnS0pNRr35BalgFmij86DaXLl68RrHQsnhZByJvnIplN+o2MwYTAP" \
+    "BgNVHRMBAf8EBTADAQH/MA4GA1UdDwEB/wQEAwICBDAdBgNVHQ4EFgQUWRCGkMIn" \
+    "dsBtfDW1emaTkWLtggYwHwYDVR0jBBgwFoAUfCtcQDvXYwkR37fHwe/mi7JyQIQw" \
+    "CgYIKoZIzj0EAwIDRwAwRAIgaK6aOuCfMwgAASavkZpoxWag49yco4d9AlxIU+rt" \
+    "U2UCIHRieWdQICIWSEHdRTXWPPEnOd7A3UmTgoqbMl+Imhy8"
+
+#define CERTIFICATE_3                                                  \
+    "MIIBnzCCAUWgAwIBAgIUe0/XdFLyc4+Sj1NMkvbagE8DFaUwCgYIKoZIzj0EAwIw" \
+    "HDELMAkGA1UEBhMCVVMxDTALBgNVBAMMBHJvb3QwIBcNMjQwMTIzMDA1NzE4WhgP" \
+    "MjIwMzA2MzAwMDU3MThaMBwxCzAJBgNVBAYTAlVTMQ0wCwYDVQQDDARyb290MFkw" \
+    "EwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEPPtDvucc4UGOdIjED2P/vDxVYDhO5P8s" \
+    "7lyys3QZpKapMuc9wOV0cQ6tN9h4kVY+FJocYgqDAl2vv6Rg/wSbl6NjMGEwHQYD" \
+    "VR0OBBYEFHwrXEA712MJEd+3x8Hv5ouyckCEMB8GA1UdIwQYMBaAFHwrXEA712MJ" \
+    "Ed+3x8Hv5ouyckCEMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/BAQDAgIEMAoG" \
+    "CCqGSM49BAMCA0gAMEUCIQCcbhKRAsqQCj2pCXh+9og3sLw9q8nU4xAB9xuV3vPA" \
+    "FAIgBxFWfRdu09dtE0IUkLTY0WPxiWaYYrKexlD4wUquqJE="
+
+#define BEGIN_CERT      "BEGIN CERTIFICATE"
+#define BEGIN_CERT_LINE "-----" BEGIN_CERT "-----"
+#define END_CERT        "END CERTIFICATE"
+#define END_CERT_LINE   "-----" END_CERT "-----"
+
+static S2N_RESULT s2n_test_validate_cert(struct s2n_cert *cert, struct s2n_blob *expected)
+{
+    RESULT_ENSURE_REF(cert);
+    RESULT_ENSURE_REF(expected);
+    RESULT_ENSURE_EQ(cert->raw.size, expected->size);
+    RESULT_ENSURE_EQ(memcmp(cert->raw.data, expected->data, expected->size), 0);
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    const char *expected_cert_strs[] = {
+        CERTIFICATE_1,
+        CERTIFICATE_2,
+        CERTIFICATE_3,
+    };
+    struct s2n_blob expected_certs[sizeof(expected_cert_strs) / sizeof(expected_cert_strs[1])] = { 0 };
+
```

Comment:
If this looks weird, that's because:
1. [s2n_array_len](https://github.com/aws/s2n-tls/blob/003412f1d6fceb12207b1c3ee4ea8bd20e0be013/utils/s2n_safety.h#L100) isn't constant, so can't be used as an array size. It probably SHOULD be constant, because we don't need a null check, because it only works for arrays, not pointers. I'm planning to fix that in another PR (it's not a one line change).
2. [grep_simple_mistakes.sh](https://github.com/aws/s2n-tls/blob/003412f1d6fceb12207b1c3ee4ea8bd20e0be013/codebuild/bin/grep_simple_mistakes.sh#L92) won't let me re-write s2n_array_len to avoid that restriction. Using [1] instead of [0] gets around that check.

I'm open to other ideas. I could just write "3", but then it wouldn't stay in sync with expected_cert_strs :(